lib/kpeg/code_generator.rb

```diff
@@ -128,129 +119,102 @@ def output_op(code, op, indent=2)
-          code << indentify("  self.pos = #{ss}\n", indent)
-          code << indentify("end\n", indent)
+          code << indentify("_tmp = true # end optional\n", indent)
         elsif op.min == 0 and !op.max
           if op.save_values
-            code << indentify("_ary = []\n", indent)
+            ary = save("_ary")
+            code << indentify("#{ary} = [] # kleene\n", indent)
+            code << indentify("while true\n", indent)
+          else
+            code << indentify("while true # kleene\n", indent)
           end
 
-          code << indentify("while true\n", indent)
           output_op code, op.op, (indent+1)
+          code << indentify("  break unless _tmp\n", indent)
           if op.save_values
-            code << indentify("  _ary << @result if _tmp\n", indent)
+            code << indentify("  #{ary} << @result\n", indent)
           end
-          code << indentify("  break unless _tmp\n", indent)
           code << indentify("end\n", indent)
-          code << indentify("_tmp = true\n", indent)
-
           if op.save_values
-            code << indentify("@result = _ary\n", indent)
+            code << indentify("@result = #{ary}\n", indent)
           end
+          code << indentify("_tmp = true # end kleene\n", indent)
 
-        elsif op.min == 1 and !op.max
-          code << indentify("#{ss} = self.pos\n", indent)
+        else
+          ss = save()
+          code << indentify("#{ss} = self.pos # repetition\n", indent)
           if op.save_values
-            code << indentify("_ary = []\n", indent)
+            ary = save("_ary")
+            incr = "#{ary} << @result"
+            cnt = "#{ary}.size"
+            code << indentify("#{ary} = []\n", indent)
+          else
+            cnt = save("_count")
+            incr = "#{cnt} += 1"
+            code << indentify("#{cnt} = 0\n", indent)
           end
-          output_op code, op.op, indent
-          code << indentify("if _tmp\n", indent)
-          if op.save_values
-            code << indentify("  _ary << @result\n", indent)
+          code << indentify("while true\n", indent)
+          output_op code, op.op, (indent+1)
+          code << indentify("  break unless _tmp\n", indent)
+          code << indentify("  #{incr}\n", indent)
+          if op.max
+            code << indentify("  break if #{cnt} == #{op.max}\n", indent)
           end
-          code << indentify("  while true\n", indent)
-          output_op code, op.op, (indent+2)
+          code << indentify("end\n", indent)
           if op.save_values
-            code << indentify("    _ary << @result if _tmp\n", indent)
+            code << indentify("@result = #{ary}\n", indent)
           end
-          code << indentify("    break unless _tmp\n", indent)
-          code << indentify("  end\n", indent)
-          code << indentify("  _tmp = true\n", indent)
+          code << indentify("_tmp = #{cnt} >= #{op.min} || nil\n", indent)
```

Comment:
Why include the `|| nil` here? _tmp being false is totally fine.

Reply:
You did it with NotPredicate: `tmp = tmp ? nil : true` . I thought falsy tmp should be nil.
Ok, will simplify things.